extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -1010,40 +1011,73 @@ export class AmpAutocomplete extends AMP.BaseElement {
   /**
    * Writes the selected value into the input field.
    * @param {?Element} element
+   * @return {?string}
    * @private
    */
-  selectItem_(element) {
+  setInputValue_(element) {
     if (element === null || element.hasAttribute('data-disabled')) {
-      return;
+      return null;
     }
     const selectedValue =
       element.getAttribute('data-value') || element.textContent;
 
-    this.fireSelectEvent_(selectedValue);
-    this.clearAllItems_();
-    this.toggleResults_(false);
-
     this.inputElement_.value = this.binding_.getUserInputForUpdateWithSelection(
       selectedValue,
       this.inputElement_,
       this.userInput_
     );
     this.userInput_ = this.binding_.getUserInputForUpdate(this.inputElement_);
+
+    return selectedValue;
+  }
+
+  /**
+   * Finish item selection with the selected value.
+   * @param {?string} value
+   * @private
+   */
+  selectItem_(value) {
+    if (value === null) {
+      return;
+    }
+    this.fireSelectEvent_(value);
+    this.clearAllItems_();
+    this.toggleResults_(false);
   }
 
   /**
-   * Triggers a 'select' event with the given value as the value emitted.
+   * Triggers select event on amp-autocomplete element and change events (bind
+   * and native, for amp-script) on child input/textarea with the given value
+   * as the value emitted.
    * @param {string} value
    * @private
    */
   fireSelectEvent_(value) {
-    const name = 'select';
-    const selectEvent = createCustomEvent(
+    let name = 'select';
+    let event = createCustomEvent(
+      this.win,
+      `amp-autocomplete.${name}`,
+      /** @type {!JsonObject} */ ({value})
+    );
+    this.action_.trigger(this.element, name, event, ActionTrust.HIGH);
+
+    // Appease type checking
+    if (!this.inputElement_) {
+      return;
+    }
+
+    // Ensure on="change" is triggered for input
+    name = 'change';
+    event = createCustomEvent(
       this.win,
       `amp-autocomplete.${name}`,
       /** @type {!JsonObject} */ ({value})
     );
-    this.action_.trigger(this.element, name, selectEvent, ActionTrust.HIGH);
+    this.action_.trigger(this.inputElement_, name, event, ActionTrust.HIGH);
+
+    // Ensure native change listeners in user amp-scripts are triggered
+    event = createCustomEvent(this.win, 'change', null);
```

Comment:
Won't this fire a second `change` event without [consuming the native one first](https://github.com/ampproject/amphtml/issues/26273#issuecomment-573090725)?

Reply:
> It is also worth considering setting a flag from the selection handler in the extension binary so the event listener can consume the native change event on the blur from the selection action, and then firing one programmatically from the selection handler with the updated value.

Could you explain in more detail how you envision consuming the native `change` event on blur? Neither `blur` nor `change` are cancelable events and I don't think we can guarantee amp-autocomplete will attach a listener early enough to stop all bubbling on `change`.

I agree multiple `change` events are often fired due to my code addition, but I haven't found a more elegant solution to: type 'a', click _orange_, type 'a', click _orange_. The second time you click on _orange_, no native `change` event is fired.

Reply:
Thanks for looking into this. I manually tested the behavior locally and it looks reasonable.